### PR TITLE
Mw/net 4888 add namespace tests failover wan fed

### PR DESCRIPTION
**Pull request by wilkermichael**
See commits for more details

Changes proposed in this PR:
- This test adds failover scenarios for Wan Federation
- Test the following scenarios:
	dc1-static-client -- nominal -- > dc1-static-server in namespace ns1
	dc1-static-client -- failover --> dc2-static-server in namespace ns1
	dc1-static-client -- failover --> dc1-static-server in namespace ns2

How I've tested this PR:
Pipeline issues out of my control, so ran the following locally for the tests connect, snapshot-agent, wan-federation:

- [x] acceptance
- [x] acceptance t-proxy
- [x] acceptance cni 

How I expect reviewers to test this PR:
👀 

Checklist:
- [ ] Tests added
- [ ] [CHANGELOG entry added](https://github.com/hashicorp/consul-k8s/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

File: acceptance/framework/connhelper/connect_helper.go [hunk 1]
```diff
@@ -225,21 +234,43 @@ func (c *ConnectHelper) TestConnectionFailureWithoutIntention(t *testing.T) {
...
+type IntentionOpts struct {
+	DestinationNamespace string
+	SourceNamespace      string
+}
+
 // CreateIntention creates an intention for the static-server pod to connect to
-// the static-client pod.
-func (c *ConnectHelper) CreateIntention(t *testing.T) {
+// the static-client pod. opts parameter allows for overriding of some fields. If opts is empty
+// then all namespaces are the KubectlOptsForApp namespaces.
```
Comment (zalimeni) [inline acceptance/framework/connhelper/connect_helper.go:244]:
```suggestion
// then all namespaces are the KubectlOptsForApp namespace.
```

File: acceptance/tests/wan-federation/wan_federation_test.go [hunk 2]
```diff
@@ -190,11 +209,259 @@ func TestWANFederation(t *testing.T) {
...
+			// Install the primary consul cluster in the default kubernetes context
+			primaryConsulCluster := consul.NewHelmCluster(t, primaryHelmValues, primaryContext, cfg, releaseName)
+			primaryConsulCluster.Create(t)
+
+			// Get the federation secret from the primary cluster and apply it to secondary cluster
+			federationSecretName := fmt.Sprintf("%s-consul-federation", releaseName)
+			logger.Logf(t, "Retrieving federation secret %s from the primary cluster and applying to the secondary", federationSecretName)
+			federationSecret, err := primaryContext.KubernetesClient(t).CoreV1().Secrets(primaryContext.KubectlOptions(t).Namespace).Get(context.Background(), federationSecretName, metav1.GetOptions{})
+			require.NoError(t, err)
+			federationSecret.ResourceVersion = ""
```
Comment (zalimeni) [inline acceptance/tests/wan-federation/wan_federation_test.go:295]: ~ Is setting this to the empty string necessary for the copy to work? Wondering whether a comment or secret copy helper would be useful here

File: acceptance/tests/wan-federation/wan_federation_test.go [hunk 3]
```diff
@@ -190,11 +209,259 @@ func TestWANFederation(t *testing.T) {
...
+			// which will not work from inside the container. That's why we need to use the endpoints address instead
+			// which will point the node IP.
+			if cfg.UseKind {
+				// The Kubernetes AuthMethod host is read from the endpoints for the Kubernetes service.
+				kubernetesEndpoint, err := secondaryContext.KubernetesClient(t).CoreV1().Endpoints("default").Get(context.Background(), "kubernetes", metav1.GetOptions{})
+				require.NoError(t, err)
+				k8sAuthMethodHost = fmt.Sprintf("%s:%d", kubernetesEndpoint.Subsets[0].Addresses[0].IP, kubernetesEndpoint.Subsets[0].Ports[0].Port)
+			} else {
+				k8sAuthMethodHost = k8s.KubernetesAPIServerHostFromOptions(t, secondaryContext.KubectlOptions(t))
+			}
```
Comment (zalimeni) [inline acceptance/tests/wan-federation/wan_federation_test.go:311]: It looks like this code has been copied in a few places but there's a nearly identical helper (adds `https://` to host) [here](https://github.com/hashicorp/consul-k8s/blob/4d40591d4eb93b07f0eb7770623cff457a082890/acceptance/framework/k8s/helpers.go#L88). Could we use that instead?

File: acceptance/tests/wan-federation/wan_federation_test.go [hunk 4]
```diff
@@ -190,11 +209,259 @@ func TestWANFederation(t *testing.T) {
...
+				Cfg:             cfg,
+				ConsulClient:    secondaryClient,
+			}
+
+			// Create Namespaces
+			primaryNamespaceOpts := primaryHelper.Ctx.KubectlOptionsForNamespace(primaryNamespace)
+			primaryHelper.SetupNamespace(t, primaryNamespaceOpts.Namespace)
+			primarySecondaryNamepsaceOpts := primaryHelper.Ctx.KubectlOptionsForNamespace(secondaryNamespace)
+			primaryHelper.SetupNamespace(t, primarySecondaryNamepsaceOpts.Namespace)
+			secondaryNamespaceOtps := secondaryHelper.Ctx.KubectlOptionsForNamespace(primaryNamespace)
```
Comment (zalimeni) [inline acceptance/tests/wan-federation/wan_federation_test.go:397]:
```suggestion
			secondaryNamespaceOpts := secondaryHelper.Ctx.KubectlOptionsForNamespace(primaryNamespace)
```

File: acceptance/framework/connhelper/connect_helper.go [hunk 5]
```diff
@@ -170,6 +175,18 @@ func (c *ConnectHelper) DeployClientAndServer(t *testing.T) {
 		})
 }
 
+func (c *ConnectHelper) SetupNamespace(t *testing.T, namespace string) {
```
Comment (curtbushko) [inline acceptance/framework/connhelper/connect_helper.go:178]: **suggestion:** I might consider naming this from SetupNamespace to CreateNamespace so that it matches the kubectl command.  (might help with searching?)

File: acceptance/tests/wan-federation/wan_federation_test.go [hunk 6]
```diff
@@ -190,11 +209,259 @@ func TestWANFederation(t *testing.T) {
...
+				UseAppNamespace: false,
+				Cfg:             cfg,
+				ConsulClient:    secondaryClient,
+			}
+
+			// Create Namespaces
+			primaryNamespaceOpts := primaryHelper.Ctx.KubectlOptionsForNamespace(primaryNamespace)
+			primaryHelper.SetupNamespace(t, primaryNamespaceOpts.Namespace)
+			primarySecondaryNamepsaceOpts := primaryHelper.Ctx.KubectlOptionsForNamespace(secondaryNamespace)
+			primaryHelper.SetupNamespace(t, primarySecondaryNamepsaceOpts.Namespace)
```
Comment (zalimeni) [inline acceptance/tests/wan-federation/wan_federation_test.go:396]: ~ It might help to briefly comment here similar to below that this is a second namespace in the primary DC... I think the "primary" and "secondary" terms here might be a bit hard to follow to someone reading the test for the first time

File: acceptance/tests/wan-federation/wan_federation_test.go [hunk 7]
```diff
@@ -190,11 +209,259 @@ func TestWANFederation(t *testing.T) {
...
+			logger.Log(t, "Creating static-client in dc1")
+			k8s.DeployKustomize(t, primaryNamespaceOpts, cfg.NoCleanupOnFailure, cfg.NoCleanup, cfg.DebugDirectory, "../fixtures/cases/wan-federation/static-client")
+
+			// Spin up a second server on dc1 in a separate namespace
+			logger.Logf(t, "Creating server on dc1 in namespace %s", primarySecondaryNamepsaceOpts.Namespace)
+			k8s.DeployKustomize(t, primarySecondaryNamepsaceOpts, cfg.NoCleanupOnFailure, cfg.NoCleanup, cfg.DebugDirectory, "../fixtures/cases/wan-federation/dc1-ns2-static-server")
+
+			// There is currently an issue that requires the intentions and resolvers to be created after
+			// the static-server/clients when using namespaces. When created before, Consul gives a "namespace does not exist"
+			// error
```
Comment (zalimeni) [inline acceptance/tests/wan-federation/wan_federation_test.go:416]: ~ Do we have a Jira/GH issue to reference for when this is fixed? NBD if not but might be useful to future-us

File: acceptance/tests/wan-federation/wan_federation_test.go [hunk 8]
```diff
@@ -8,17 +8,36 @@ import (
...
 	"github.com/hashicorp/consul-k8s/acceptance/framework/logger"
+	"github.com/hashicorp/consul/sdk/testutil/retry"
+	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-const StaticClientName = "static-client"
+const (
+	StaticClientName = "static-client"
```
Comment (curtbushko) [inline acceptance/tests/wan-federation/wan_federation_test.go:26]: **nitpick:** Should probably be StaticClientName with a lowercase 's' (though you didn't do it).

File: acceptance/tests/wan-federation/wan_federation_test.go [hunk 9]
```diff
@@ -190,11 +209,259 @@ func TestWANFederation(t *testing.T) {
...
+
+			// There is currently an issue that requires the intentions and resolvers to be created after
+			// the static-server/clients when using namespaces. When created before, Consul gives a "namespace does not exist"
+			// error
+			if c.secure {
+				// Only need to create intentions in the primary datacenter as they will be replicated to the secondary
+				// ns1 static-client (source) -> ns1 static-server (destination)
+				primaryHelper.CreateIntention(t, connhelper.IntentionOpts{DestinationNamespace: primaryNamespaceOpts.Namespace, SourceNamespace: primaryNamespaceOpts.Namespace})
+
+				// ns1 static-client (source) -> ns2 static-sever (destination)
```
Comment (zalimeni) [inline acceptance/tests/wan-federation/wan_federation_test.go:422]:
```suggestion
				// ns1 static-client (source) -> ns2 static-server (destination)
```

**Review by curtbushko — APPROVED** [review]
:chefskiss: Looks amazing!

I have some suggestions but nothing I would necessarily block on.

File: acceptance/tests/wan-federation/wan_federation_test.go [hunk 10]
```diff
@@ -190,11 +209,259 @@ func TestWANFederation(t *testing.T) {
...
+}
+
+// serviceFailoverCheck verifies that the server failed over as expected by checking that curling the `static-server`
+// using the `static-client` responds with the expected cluster name. Each static-server responds with a unique
+// name so that we can verify failover occurred as expected.
+func serviceFailoverCheck(t *testing.T, options *terratestK8s.KubectlOptions, port string, expectedName string) {
+	timer := &retry.Timer{Timeout: retryTimeout, Wait: 5 * time.Second}
+	var resp string
+	var err error
+	retry.RunWith(timer, t, func(r *retry.R) {
```
Comment (zalimeni) [inline acceptance/tests/wan-federation/wan_federation_test.go:460]: FWIW, the locality-aware routing test I just added in `consul` had a false positive when retries were used (random LB eventually returned the correct name).

I'm pretty sure that doesn't apply here based on my understanding of how `Failover` works (all or nothing), but if that's not the case, we could do a non-retried "make a few extra calls" loop after this one to verify the host is consistent.

Again, probably not necessary here, but thought I'd mention since it was something that tripped another test up recently.

**Review by zalimeni — APPROVED** [review]
One question about `SourceNamespace ` vs. `DestinationNamespace`, other comments are minor and generally non-blockers. Generally LGTM!

Approving to unblock once those are resolved, though wouldn't mind a second set of 👀 from @curtbushko if he has time.

File: acceptance/tests/wan-federation/wan_federation_test.go (hunk 8 above)
Comment (wilkermichael) [reply acceptance/tests/wan-federation/wan_federation_test.go:26]: got rid of it, using the `connhelper` imported variable instead which is the same

File: acceptance/framework/connhelper/connect_helper.go [hunk 11]
```diff
@@ -170,6 +175,18 @@ func (c *ConnectHelper) DeployClientAndServer(t *testing.T) {
...
+	opts := c.KubectlOptsForApp(t)
```
[the remark this answers is not in the document]
Comment (wilkermichael) [reply acceptance/framework/connhelper/connect_helper.go:179]: No, I don't think so

File: acceptance/tests/wan-federation/wan_federation_test.go (hunk 2 above)
Comment (wilkermichael) [reply acceptance/tests/wan-federation/wan_federation_test.go:295]: Not sure, I copy/pasted it from the test above

File: acceptance/tests/wan-federation/wan_federation_test.go (hunk 10 above)
Comment (wilkermichael) [reply acceptance/tests/wan-federation/wan_federation_test.go:460]: that would happen if I wasn't mirroring the namespaces, I can try adding the second check after the retries

File: acceptance/tests/wan-federation/wan_federation_test.go (hunk 7 above)
Comment (wilkermichael) [reply acceptance/tests/wan-federation/wan_federation_test.go:416]: No we don't, I don't know if we create github issues? and I don't want to link a jira issue in a public repo. I'll create a jira issue for it.

File: acceptance/tests/wan-federation/wan_federation_test.go (hunk 3 above)
Comment (wilkermichael) [reply acceptance/tests/wan-federation/wan_federation_test.go:311]: I'll try it out, not sure if `https://` might have a negative effect on the test

File: acceptance/tests/wan-federation/wan_federation_test.go (hunk 2 above)
Comment (wilkermichael) [reply acceptance/tests/wan-federation/wan_federation_test.go]: I'll create a copy helper to follow DRY a bit more

File: acceptance/tests/wan-federation/wan_federation_test.go (hunk 3 above)
Comment (wilkermichael) [reply acceptance/tests/wan-federation/wan_federation_test.go]: looks like it works 🚀 !

File: acceptance/framework/connhelper/connect_helper.go [hunk 12]
```diff
@@ -300,15 +307,17 @@ func (c *ConnectHelper) CreateResolverRedirect(t *testing.T) {
...
+// server fails when no intentions are configured. When provided with a ClientType option
+// the client is overridden, otherwise a default will be used
+func (c *ConnectHelper) TestConnectionFailureWithoutIntention(t *testing.T, connHelperOpts ConnHelperOpts) {
 	logger.Log(t, "checking that the connection is not successful because there's no intention")
 	opts := c.KubectlOptsForApp(t)
 	//Default to deploying static-client. If a client type is passed in (ex. job-client), use that instead.
 	client := StaticClientName
-	if len(clientType) > 0 {
-		client = clientType[0]
+	if connHelperOpts.ClientType != "" {
```
Comment (curtbushko) [inline acceptance/framework/connhelper/connect_helper.go:317]: **suggestion:** All the way through these I would named it connHelperOpts -> opts.

Comment (wilkermichael) [reply acceptance/framework/connhelper/connect_helper.go:317]: Discussed with Curt, will leave it because opts is used in most of the functions for the KubectlOptions
